### DISCUSSION TITLE
Provide credentials to https repo url

### DISCUSSION THREAD

**Discussion by bmorris591** [GitoxideLabs/gitoxide#772] [Q&A] [answered]
I am trying to use the `gix` crate to clone and push to a gitlab repo, I want to use the https URL as it allows auth with a personal access token.

I am looking for a way to supply credentials - I working from the [clone example][1].

Using the `git` CLI you can pass username&password in the URL as the userinfo, e.g. `http://me:password@host.com/...` but looking at the code for `gix_url::Url` it doesn't handle that part of the URL and discards it.

So far I've seen that `remote::connection::Connection` has `authenticate` - but that is set to `None` in `to_connection_with_transport` and I don't see a way to change that.

The there is also `connection::with_credentials` - but that doesn't seem to be usage without using a custom `Connection` - as opposed to the API in the clone example.

So far I've hacked it by
- creating a file with a store formatted line for the repo 
- create a `open::Options` and supply `config_overrides` to ``credential.helper=store --file {}`
- using `clone::PrepareFetch::new` (rather than `gix::prepare_clone`) and the options

This seems more hacky than I'd like - it would be neater (at least to me) to be able to pass credentials in programmatically. Is that possible?

[1]: https://github.com/Byron/gitoxide/blob/main/gix/examples/clone.rs

**Comment by Byron** [comment, marked as answer]
Thanks for bringing this up!

When implementing `gix-url` I felt strongly about not dealing with passwords directly knowing very well that such cases exists like the one presented here. Now I think this stance was way too opinionated and compatibility with `git` is more important than that.

Let me see if I can make a password field available ([see PR](https://github.com/Byron/gitoxide/pull/774)).

**Reply by Byron** [reply]
The PR has landed in `main` and I would expect these URLs to work out of the box now, please let me know if you encounter any other problems though.
